linkerd/app/outbound/src/http/logical/profile.rs

```diff
@@ -0,0 +1,308 @@
+use super::{
+    super::{concrete, retry},
+    BackendCache, Concrete, Distribution, NoRoute,
+};
+use linkerd_app_core::{
+    classify, metrics,
+    proxy::http::{self, balance},
+    svc, Error,
+};
+use linkerd_distribute as distribute;
+use std::{fmt::Debug, hash::Hash, sync::Arc, time};
+
+pub use linkerd_app_core::profiles::{
+    http::{route_for_request, RequestMatch, Route},
+    LogicalAddr, Profile, Target,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Routes {
+    pub addr: LogicalAddr,
+    pub routes: Arc<[(RequestMatch, Route)]>,
+    pub targets: Arc<[Target]>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct Params<T: Clone + Debug + Eq + Hash> {
+    parent: T,
+    addr: LogicalAddr,
+    profile_routes: Arc<[(RequestMatch, RouteParams<T>)]>,
+    backends: distribute::Backends<Concrete<T>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(super) struct RouteParams<T> {
+    parent: T,
+    addr: LogicalAddr,
+    profile: Route,
+    distribution: Distribution<T>,
+}
+
+// === impl Params ===
+
+// Wraps a `NewService`--instantiated once per logical target--that caches a set
+// of concrete services so that, as the watch provides new `Params`, we can
+// reuse inner services.
+impl<T> Params<T>
+where
+    T: Clone + Debug + Eq + Hash + Send + Sync + 'static,
+{
```

Comment:
tiny nit: was this intended to be a doc comment on `layer`?
```suggestion
impl<T> Params<T>
where
    T: Clone + Debug + Eq + Hash + Send + Sync + 'static,
{
    /// Wraps a `NewService`--instantiated once per logical target--that caches
    /// a set of concrete services so that, as the watch provides new `Params`,
    /// we can reuse inner services.
```